core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,164 @@
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    deleteFiles(manifestPathsToDelete, "manifest");
+    deleteFiles(manifestListsToDelete, "manifest list");
+  }
+
+  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
+    Set<ManifestFile> manifestFiles = Sets.newHashSet();
+    for (Snapshot snapshot : snapshots) {
+      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
+        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
+          manifestFiles.add(manifestFile.copy());
+        }
+      } catch (IOException e) {
+        throw new RuntimeIOException(
+            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
+      }
+    }
+
+    return manifestFiles;
+  }
+
+  // Helper to determine data files to delete
+  private Set<String> findFilesToDelete(
+      Set<ManifestFile> manifestFilesToDelete, Set<ManifestFile> currentManifestFiles) {
+    Set<String> filesToDelete = ConcurrentHashMap.newKeySet();
+
+    Tasks.foreach(manifestFilesToDelete)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn(
+                    "Failed to determine live files in manifest {}: this may cause orphaned data files",
+                    item.path(),
+                    exc))
+        .run(
+            manifest -> {
+              try (CloseableIterable<String> paths = ManifestFiles.readPaths(manifest, fileIO)) {
+                paths.forEach(filesToDelete::add);
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+
+    if (filesToDelete.isEmpty()) {
+      return filesToDelete;
+    }
+
+    try {
+      Tasks.foreach(currentManifestFiles)
+          .retry(3)
+          .stopOnFailure()
+          .throwFailureWhenFinished()
+          .executeWith(planExecutorService)
+          .onFailure(
+              (item, exc) ->
+                  LOG.warn(
+                      "Failed to determine live files in manifest {}: this may cause orphaned data files",
+                      item.path(),
+                      exc))
+          .run(
+              manifest -> {
+                if (filesToDelete.isEmpty()) {
+                  return;
+                }
+
+                // Remove all the live files from the candidate deletion set
+                try (CloseableIterable<String> paths = ManifestFiles.readPaths(manifest, fileIO)) {
+                  paths.forEach(filesToDelete::remove);
+                } catch (IOException e) {
+                  throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+                }
+              });
+
+    } catch (Throwable e) {
+      LOG.warn("Failed to determine the data files to be removed", e);
```

Comment:
As a follow up, I think this message could be a little better. This makes sense if you know what the code is currently doing, but if you don't then I think it will be hard to understand that the error was in reading current manifests, rather than dropped manifests. I'd use something like "Failed to list all reachable files".